intermediate_notebooks/E2E/taxi/NYCTaxi-E2E-GPU-Dask.ipynb

```diff
@@ -0,0 +1,1995 @@
+       "    </tr>\n",
+       "    <tr>\n",
+       "      <th>643185</th>\n",
+       "      <td>2016-01-02 23:46:41</td>\n",
+       "      <td>2016-01-03 00:39:11</td>\n",
+       "      <td>1</td>\n",
+       "      <td>51.700001</td>\n",
+       "      <td>-73.837006</td>\n",
+       "      <td>40.753101</td>\n",
+       "      <td>5</td>\n",
+       "      <td>-73.210365</td>\n",
+       "      <td>41.183952</td>\n",
+       "      <td>0.0</td>\n",
+       "    </tr>\n",
+       "  </tbody>\n",
+       "</table>\n",
+       "</div>"
+      ],
+      "text/plain": [
+       "           pickup_datetime    dropoff_datetime  passenger_count  \\\n",
+       "156952 2016-01-01 04:11:50 2016-01-01 05:16:12                1   \n",
+       "202906 2016-01-01 10:30:44 2016-01-01 11:05:51                1   \n",
+       "500926 2016-01-02 14:12:17 2016-01-02 16:04:43                1   \n",
+       "528510 2016-01-02 15:43:24 2016-01-02 16:30:11                1   \n",
+       "643185 2016-01-02 23:46:41 2016-01-03 00:39:11                1   \n",
+       "\n",
+       "        trip_distance  pickup_longitude  pickup_latitude  rate_code  \\\n",
+       "156952      53.400002        -73.940117        40.709667          1   \n",
+       "202906     175.199997        -73.790001        40.644051          1   \n",
+       "500926      63.900002        -73.806465        40.685558          5   \n",
+       "528510     208.100006        -73.967690        40.802883          1   \n",
+       "643185      51.700001        -73.837006        40.753101          5   \n",
+       "\n",
+       "        dropoff_longitude  dropoff_latitude  fare_amount  \n",
+       "156952         -74.017479         40.635601          2.5  \n",
+       "202906         -73.953697         40.766068          2.5  \n",
+       "500926         -74.574554         40.366108          0.0  \n",
+       "528510         -73.783188         40.643898          2.5  \n",
+       "643185         -73.210365         41.183952          0.0  "
+      ]
+     },
+     "execution_count": 11,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "# check out if there is any abnormal data where trip distance is long, but the fare is very low.\n",
+    "taxi_df[(taxi_df.trip_distance > 50) & (taxi_df.fare_amount < 50)].head()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 12,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#2016-01 data is used for visuals.\n",
+    "taxi_df_cdf = clean(dask_cudf.read_csv(valid_files[0])).compute()\n",
+    "taxi_df_pd = taxi_df_cdf.to_pandas()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The plot below visualizes the histogram of trip_distance and we can see some abnormal trip_distance values for some records. Taking this and also the NYC map coordinates into consideration, we will only select records where tripdistance < 500 miles."
```

Comment:
Pulling all 2016 data to the client is slow and OOMs when working on the full dataset. Using Pandas to compute the histogram is also slow.

I was looking at a way to compute the trip distance buckets on device and just visualize the results:
```
taxi_df.trip_distance.describe().compute()

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-11-a6454e4baf74> in <module>
----> 1 taxi_df.trip_distance.quantile().compute()

~/conda/envs/rapids-20200508_1351/lib/python3.7/site-packages/dask/base.py in compute(self, **kwargs)
    164         dask.base.compute
    165         """
--> 166         (result,) = compute(self, traverse=False, **kwargs)
    167         return result
    168 

~/conda/envs/rapids-20200508_1351/lib/python3.7/site-packages/dask/base.py in compute(*args, **kwargs)
    435     keys = [x.__dask_keys__() for x in collections]
    436     postcomputes = [x.__dask_postcompute__() for x in collections]
--> 437     results = schedule(dsk, keys, **kwargs)
    438     return repack([f(r, *a) for r, (f, a) in zip(results, postcomputes)])
    439 

~/conda/envs/rapids-20200508_1351/lib/python3.7/site-packages/distributed/client.py in get(self, dsk, keys, restrictions, loose_restrictions, resources, sync, asynchronous, direct, retries, priority, fifo_timeout, actors, **kwargs)
   2664                     should_rejoin = False
   2665             try:
-> 2666                 results = self.gather(packed, asynchronous=asynchronous, direct=direct)
   2667             finally:
   2668                 for f in futures.values():

~/conda/envs/rapids-20200508_1351/lib/python3.7/site-packages/distributed/client.py in gather(self, futures, errors, direct, asynchronous)
   1965                 direct=direct,
   1966                 local_worker=local_worker,
-> 1967                 asynchronous=asynchronous,
   1968             )
   1969 

~/conda/envs/rapids-20200508_1351/lib/python3.7/site-packages/distributed/client.py in sync(self, func, asynchronous, callback_timeout, *args, **kwargs)
    814         else:
    815             return sync(
--> 816                 self.loop, func, *args, callback_timeout=callback_timeout, **kwargs
    817             )
    818 

~/conda/envs/rapids-20200508_1351/lib/python3.7/site-packages/distributed/utils.py in sync(loop, func, callback_timeout, *args, **kwargs)
    345     if error[0]:
    346         typ, exc, tb = error[0]
--> 347         raise exc.with_traceback(tb)
    348     else:
    349         return result[0]

~/conda/envs/rapids-20200508_1351/lib/python3.7/site-packages/distributed/utils.py in f()
    329             if callback_timeout is not None:
    330                 future = asyncio.wait_for(future, callback_timeout)
--> 331             result[0] = yield future
    332         except Exception as exc:
    333             error[0] = sys.exc_info()

~/conda/envs/rapids-20200508_1351/lib/python3.7/site-packages/tornado/gen.py in run(self)
    733 
    734                     try:
--> 735                         value = future.result()
    736                     except Exception:
    737                         exc_info = sys.exc_info()

~/conda/envs/rapids-20200508_1351/lib/python3.7/site-packages/distributed/client.py in _gather(self, futures, errors, direct, local_worker)
   1824                             exc = CancelledError(key)
   1825                         else:
-> 1826                             raise exception.with_traceback(traceback)
   1827                         raise exc
   1828                     if errors == "skip":

~/conda/envs/rapids-20200508_1351/lib/python3.7/site-packages/dask/optimization.py in __call__()
    989         if not len(args) == len(self.inkeys):
    990             raise ValueError("Expected %d args, got %d" % (len(self.inkeys), len(args)))
--> 991         return core.get(self.dsk, self.outkey, dict(zip(self.inkeys, args)))
    992 
    993     def __reduce__(self):

~/conda/envs/rapids-20200508_1351/lib/python3.7/site-packages/dask/core.py in get()
    149     for key in toposort(dsk):
    150         task = dsk[key]
--> 151         result = _execute_task(task, cache)
    152         cache[key] = result
    153     result = _execute_task(out, cache)

~/conda/envs/rapids-20200508_1351/lib/python3.7/site-packages/dask/core.py in _execute_task()
    119         # temporaries by their reference count and can execute certain
    120         # operations in-place.
--> 121         return func(*(_execute_task(a, cache) for a in args))
    122     elif not ishashable(arg):
    123         return arg

~/conda/envs/rapids-20200508_1351/lib/python3.7/site-packages/dask/core.py in <genexpr>()
    119         # temporaries by their reference count and can execute certain
    120         # operations in-place.
--> 121         return func(*(_execute_task(a, cache) for a in args))
    122     elif not ishashable(arg):
    123         return arg

~/conda/envs/rapids-20200508_1351/lib/python3.7/site-packages/dask/core.py in _execute_task()
    119         # temporaries by their reference count and can execute certain
    120         # operations in-place.
--> 121         return func(*(_execute_task(a, cache) for a in args))
    122     elif not ishashable(arg):
    123         return arg

~/conda/envs/rapids-20200508_1351/lib/python3.7/site-packages/dask/optimization.py in __call__()
    989         if not len(args) == len(self.inkeys):
    990             raise ValueError("Expected %d args, got %d" % (len(self.inkeys), len(args)))
--> 991         return core.get(self.dsk, self.outkey, dict(zip(self.inkeys, args)))
    992 
    993     def __reduce__(self):

~/conda/envs/rapids-20200508_1351/lib/python3.7/site-packages/dask/core.py in get()
    149     for key in toposort(dsk):
    150         task = dsk[key]
--> 151         result = _execute_task(task, cache)
    152         cache[key] = result
    153     result = _execute_task(out, cache)

~/conda/envs/rapids-20200508_1351/lib/python3.7/site-packages/dask/core.py in _execute_task()
    119         # temporaries by their reference count and can execute certain
    120         # operations in-place.
--> 121         return func(*(_execute_task(a, cache) for a in args))
    122     elif not ishashable(arg):
    123         return arg

~/conda/envs/rapids-20200508_1351/lib/python3.7/site-packages/dask/utils.py in apply()
     28 def apply(func, args, kwargs=None):
     29     if kwargs:
---> 30         return func(*args, **kwargs)
     31     else:
     32         return func(*args)

~/conda/envs/rapids-20200508_1351/lib/python3.7/site-packages/dask/dataframe/core.py in apply_and_enforce()
   5135             return meta
   5136         if is_dataframe_like(df):
-> 5137             check_matching_columns(meta, df)
   5138             c = meta.columns
   5139         else:

~/conda/envs/rapids-20200508_1351/lib/python3.7/site-packages/dask/dataframe/utils.py in check_matching_columns()
    679             extra_info = "Order of columns does not match"
    680         raise ValueError(
--> 681             "The columns in the computed data do not match"
    682             " the columns in the provided metadata\n"
    683             f"{extra_info}"

ValueError: The columns in the computed data do not match the columns in the provided metadata
  Extra:   ['RatecodeID']
  Missing: ['ratecodeid']
```

It looks like the `clean` logic isn't handling `ratecodeid` correctly.